File: 0078-subsets/0078-subsets.cpp
class Solution {
public:
    void subset(vector<vector<int>>&ans,vector<int>&temp,vector<int>& nums,int i){
        
        if(i == nums.size()){
            ans.push_back(temp);
            return;
        }
        
        temp.push_back(nums[i]);
        //Include
        subset(ans,temp,nums,i+1);

        //Backtracking part
        temp.pop_back();
        //Exclude
        subset(ans,temp,nums,i+1);

    }
    vector<vector<int>> subsets(vector<int>& nums) {
         vector<vector<int>>ans;vector<int>temp;int i = 0;
         subset(ans,temp,nums,i);
         return ans;
    }
};

File: 0078-subsets/0078-subsets.java
class Solution {

    void help(int idx , int[] nums,  List<List<Integer>> ans,List<Integer>temp){
        if(idx == nums.length){
            ans.add(new ArrayList<>(temp));
            return;
        }
        if(idx >= nums.length) return;

        temp.add(nums[idx]);
        help(idx+1, nums,ans ,temp);
        temp.remove(temp.size()-1);
        help(idx+1, nums,ans ,temp);
    }
    public List<List<Integer>> subsets(int[] nums) {
        List<Integer>temp = new ArrayList<>();
        List<List<Integer>>ans =  new ArrayList<>();
        help(0, nums, ans,temp);
        return ans;
    }
}